sql/decimal.go

```diff
@@ -0,0 +1,216 @@
+func MustCreateDecimalType(precision uint8, scale uint8) DecimalType {
+	dt, err := CreateDecimalType(precision, scale)
+	if err != nil {
+		panic(err)
+	}
+	return dt
+}
+
+// Type implements Type interface.
+func (t decimalType) Type() query.Type {
+	return sqltypes.Decimal
+}
+
+// Compare implements Type interface.
+func (t decimalType) Compare(a interface{}, b interface{}) (int, error) {
+	if hasNulls, res := compareNulls(a, b); hasNulls {
+		return res, nil
+	}
+
+	af, err := t.ConvertToDecimal(a)
+	if err != nil {
+		return 0, err
+	}
+	bf, err := t.ConvertToDecimal(b)
+	if err != nil {
+		return 0, err
+	}
+
+	return af.Decimal.Cmp(bf.Decimal), nil
+}
+
+// Convert implements Type interface.
+func (t decimalType) Convert(v interface{}) (interface{}, error) {
+	dec, err := t.ConvertToDecimal(v)
+	if err != nil {
+		return nil, err
+	}
+	if !dec.Valid {
+		return nil, nil
+	}
+	return dec.Decimal.StringFixed(int32(t.scale)), nil
+}
+
+// Precision returns the precision, or total number of digits, that may be held.
+func (t decimalType) ConvertToDecimal(v interface{}) (decimal.NullDecimal, error) {
+	if v == nil {
+		return decimal.NullDecimal{}, nil
+	}
+
+	var res decimal.Decimal
+
+	switch value := v.(type) {
+	case int:
+		return t.ConvertToDecimal(int64(value))
+	case uint:
+		return t.ConvertToDecimal(uint64(value))
+	case int8:
+		return t.ConvertToDecimal(int64(value))
+	case uint8:
+		return t.ConvertToDecimal(uint64(value))
+	case int16:
+		return t.ConvertToDecimal(int64(value))
+	case uint16:
+		return t.ConvertToDecimal(uint64(value))
+	case int32:
+		res = decimal.NewFromInt32(value)
+	case uint32:
+		return t.ConvertToDecimal(uint64(value))
+	case int64:
+		res = decimal.NewFromInt(value)
+	case uint64:
+		res = decimal.NewFromBigInt(new(big.Int).SetUint64(value), 0)
+	case float32:
+		res = decimal.NewFromFloat32(value)
+	case float64:
+		res = decimal.NewFromFloat(value)
+	case string:
+		var err error
+		res, err = decimal.NewFromString(value)
+		if err != nil {
+			// The decimal library cannot handle all of the different formats
+			bf, _, err := new(big.Float).SetPrec(217).Parse(value, 0)
+			if err != nil {
+				return decimal.NullDecimal{}, err
+			}
+			res, err = decimal.NewFromString(bf.Text('f', -1))
+			if err != nil {
+				return decimal.NullDecimal{}, err
+			}
+		}
+	case *big.Float:
+		return t.ConvertToDecimal(value.Text('f', -1))
+	case *big.Int:
+		return t.ConvertToDecimal(value.Text(10))
+	case *big.Rat:
+		return t.ConvertToDecimal(new(big.Float).SetRat(value))
+	case decimal.Decimal:
+		res = value
+	case decimal.NullDecimal:
+		// This is the equivalent of passing in a nil
+		if !value.Valid {
+			return decimal.NullDecimal{}, nil
+		}
+		res = value.Decimal
+	default:
+		return decimal.NullDecimal{}, ErrConvertingToDecimal.New(v)
+	}
+
+	res = res.Round(int32(t.scale))
+	max, _ := decimal.NewFromString("1" + strings.Repeat("0", int(t.precision - t.scale)))
```

Comment:
So it took a second to figure out if / why this was right, which is fine, but is a sign this could use a comment (or better yet, a method. Call it maxValuePlusOne or something)

For this calculation, you can save quite a few cycles for the majority of cases  (precision < 31) if you use decimal.New(1, precision-scale) here instead.